app/src/main/java/org/mozilla/reference/browser/tabs/SyncedTabsFragment.kt

```diff
@@ -0,0 +1,310 @@
+        updateUI()
+
+        refreshLayout.setOnRefreshListener {
+            scope.launch {
+                accountManager.withConstellation {
+                    it.refreshDevicesAsync().await()
+                }
+                accountManager.syncNowAsync(SyncReason.User)
+            }
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        // Kick off a sync in the background so we get the freshest tabs.
+        scope.launch {
+            accountManager.syncNowAsync(SyncReason.User).await()
+        }
+    }
+
+    fun updateUI() {
+        val accountManager = requireComponents.backgroundServices.accountManager
+        val account = accountManager.authenticatedAccount()
+
+        when {
+            // Signed-in, no problems.
+            account != null && !accountManager.accountNeedsReauth() -> {
+                if (SyncEnginesStorage(requireContext()).getStatus()[SyncEngine.Tabs] == true) {
+                    // Tabs engine enabled.
+                    showTabs()
+                } else {
+                    // Tab engine disabled
+                    showInfoMessage("Please enable tab syncing.")
+                }
+            }
+            // Signed-in, need to re-authenticate.
+            account != null && accountManager.accountNeedsReauth() -> {
+                showInfoMessage("Please re-authenticate.")
+            }
+            // Signed-out.
+            else -> {
+                showInfoMessage("Connect with a Firefox Account.")
+            }
+        }
+    }
+
+    private fun showInfoMessage(msg: String) {
+        synced_tabs_status.text = msg
+        synced_tabs_list.visibility = View.GONE
+        synced_tabs_status.visibility = View.VISIBLE
+    }
+
+    private fun showTabs() {
+        synced_tabs_list.visibility = View.VISIBLE
+        synced_tabs_status.visibility = View.GONE
+        scope.launch {
+            val syncedTabs = syncedTabsFeature.getSyncedTabs()
+            scope.launch(Dispatchers.Main) {
+                if (syncedTabs.isEmpty() &&
+                    accountManager.authenticatedAccount()?.deviceConstellation()?.state()?.otherDevices?.isEmpty()
+                    == true) {
+                    showInfoMessage("Connect another device.")
+                } else {
+                    val items = ArrayList<DeviceOrTabHolder>()
+                    syncedTabs.forEach { (device, tabs) ->
+                        if (tabs.isNotEmpty()) {
+                            items.add(DeviceOrTabHolder.DeviceHolder(device))
+                            tabs.forEach { tab ->
+                                items.add(DeviceOrTabHolder.TabEntryHolder(tab.active()))
+                            }
+                        }
+                    }
+                    syncedTabsAdapter.submitList(items)
+                }
+            }
+        }
+    }
+
+    private class SyncedTabsAccountObserver(private val inner: SyncedTabsFragment) : AccountObserver {
```

Comment:
Can we refactor these to _not_ pass in the fragment, but instead a presenter? The presenter then invokes these methods.